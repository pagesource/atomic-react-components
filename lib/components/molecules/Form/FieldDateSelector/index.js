// @flow
export { default } from './FieldDateSelector';
export { FieldDateSelectorVanilla } from './FieldDateSelector';
