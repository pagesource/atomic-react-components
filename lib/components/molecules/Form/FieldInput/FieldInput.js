// @flow
import React from 'react';
import type { Node, ComponentType } from 'react';
import { Field, ErrorMessage } from 'formik';
import uniqueId from 'lodash/uniqueId';
import styled from 'styled-components';
import Input from '../../../atoms/Input';
import FieldError from '../../../atoms/FieldError';
import Label from '../../../atoms/Label';
import styles from './FieldInput.style';
import Para from '../../../atoms/Para';

type Props = {
  id: string,
  className?: string,
  type?: string,
  name: string,
  label: Node,
  labelProps?: Object,
  fieldProps?: Object,
  value?: string | number,
  placeholder?: string,
  validationInstructions?: string,
  onChange?: Function,
  onBlur?: Function,
};

const FieldInput = ({
  id,
  className,
  type,
  name,
  label,
  onChange,
  onBlur,
  labelProps,
  fieldProps,
  validationInstructions,
  ...others
}: Props): Node => {
  const inputID = id || uniqueId('input-');
  return (
    // TODO: Add support for label position
    <div className={className}>
      <Label htmlFor={id} {...labelProps}>
        {label}
      </Label>
      <Field id={id} name={name} type={type} {...fieldProps}>
        {({ field, form }) => {
          const { handleChange, handleBlur } = form;
          return (
            <Input
              type={type}
              {...others}
              {...field}
              aria-describedby={validationInstructions ? `${inputID}-instruction` : null}
              onChange={(e) => {
                handleChange(e);
                if (typeof onChange === 'function') onChange(e, form);
              }}
              onBlur={(e) => {
                handleBlur(e);
                if (typeof onBlur === 'function') onBlur(e, form);
              }}
            />
          );
        }}
      </Field>
      {validationInstructions ? (
        <Para id={`${inputID}-instruction`}>{validationInstructions}</Para>
      ) : (
        ''
      )}
      <ErrorMessage name={name} component={FieldError} />
    </div>
  );
};

FieldInput.defaultProps = {
  placeholder: '',
  validationInstructions: '',
  className: '',
  type: 'text',
  labelProps: {},
  value: '',
  fieldProps: {},
  onChange: null,
  onBlur: null,
};

const StyledFieldInput: ComponentType<Props> = styled(FieldInput)`
  ${styles};
`;

export default StyledFieldInput;
export { FieldInput as FieldInputVanilla };
