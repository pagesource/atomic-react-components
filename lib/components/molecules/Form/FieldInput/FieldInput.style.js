import { css } from 'styled-components';

const styles = css`
  margin-bottom: ${(props) => props.theme.spacing.UNIT_1};

  label,
  input {
    display: block;
    margin-bottom: ${(props) => props.theme.spacing.HALF_UNIT};
  }
`;

export default styles;
