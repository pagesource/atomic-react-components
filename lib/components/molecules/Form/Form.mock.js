import { action } from '@storybook/addon-actions';

const userFormProps = {
  initialValues: {
    email: '',
    username: '',
    typeOfUser: 'Admin',
    comment: '',
    ageGroup: '',
    termsAndConditions: false,
  },
  debug: false,
  action: '/',
  validationSchema: {
    username: {
      max: [15, 'Max limit reached'],
      required: 'User name Required',
    },
    email: {
      max: [30, 'Max reached'],
      email: 'Invalid email format',
      required: 'Required',
    },
    comment: {
      max: [100, 'Comment should not be more than 100 characters'],
      required: 'Please enter a comment',
    },
    ageGroup: {
      required: 'Please select an Age group',
    },
    termsAndConditions: {
      type: 'bool',
      test: [
        'termsAndConditions',
        'please accept terms and conditions!!!',
        (value) => value === true,
      ],
    },
  },
  onSubmit: action('On Submit Fired'),
};

const userFormSelectProps = {
  options: ['Admin', 'Read Only'],
  name: 'typeOfUser',
  label: 'Type of user',
};

const userFormNameInputProps = {
  name: 'username',
  type: 'text',
  label: 'User Name',
  validationInstructions: 'Enter valid username',
};

const fileMultipleInputProps = {
  name: 'fileuploadMulti',
  localeText: {
    btnTextPrimary: 'Choose Files',
    btnTextSecondary: 'Upload...',
    dragHelperText: 'Drop Files Here...',
  },
  multiple: true,
};
const fileSingleInputProps = {
  name: 'fileupload',
  localeText: {
    btnTextPrimary: 'Choose File',
    btnTextSecondary: 'Upload...',
    dragHelperText: 'Drop File Here...',
  },
};
const fileDDInputProps = {
  name: 'fileuploadDD',
  localeText: {
    btnTextPrimary: 'Choose File',
    btnTextSecondary: 'Upload...',
    dragHelperText: 'Drop File Here...',
  },
};

const userFormEmailProps = {
  type: 'email',
  name: 'email',
  label: 'Email',
  validationInstructions: 'Enter valid email',
};

const userFormCommentProps = {
  name: 'comment',
  label: 'Comment',
  placeholder: 'Enter your comment here...',
};

const userFormAgeGroupProps = {
  name: 'ageGroup',
  radioGroupArray: [
    {
      value: '30-',
      label: '< 30',
    },
    {
      value: '30-60',
      label: '30-60',
    },
    {
      value: '60+',
      label: '> 60',
    },
  ],
};

const userFormTnCProps = {
  name: 'termsAndConditions',
  label: 'I accept terms & conditions',
};

const userFormToggleProps = {
  name: 'test_1',
  id: 'checkbox_test_1',
  value: 'off',
  label: 'Toggle',
  indicatorOn: 'On',
  indicatorOff: 'Off',
  disabled: false,
  intialToggleState: false,
  displayIndicators: true,
};

export {
  userFormProps,
  fileMultipleInputProps,
  fileSingleInputProps,
  userFormSelectProps,
  userFormNameInputProps,
  userFormEmailProps,
  userFormCommentProps,
  userFormAgeGroupProps,
  userFormTnCProps,
  userFormToggleProps,
  fileDDInputProps,
};
