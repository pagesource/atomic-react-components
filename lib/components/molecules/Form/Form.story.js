// @flow
import React from 'react';
import { storiesOf } from '@storybook/react';
import {
  userFormProps,
  userFormSelectProps,
  fileMultipleInputProps,
  fileSingleInputProps,
  userFormNameInputProps,
  userFormEmailProps,
  userFormCommentProps,
  userFormAgeGroupProps,
  userFormDateSelectorProps,
  userFormTnCProps,
  userFormToggleProps,
} from './Form.mock';

// Import Vanilla Component to display markup and props
import Form from './Form';
import Button from '../../atoms/Button';
// Import Styled Component to showcase variations

storiesOf('Molecules/Form', module)
  .addParameters({ jest: ['Form', 'Button'] })
  .add('Default', () => (
    <Form {...userFormProps}>
      <div>
        <Form.Select {...userFormSelectProps} />
        <Form.Input {...userFormNameInputProps} />
        <Form.Input {...userFormEmailProps} />
        <Form.FieldFile {...fileMultipleInputProps} />
        <Form.FieldFile {...fileSingleInputProps} />
        <Form.DDFile {...fileSingleInputProps} />
        <Form.RadioGroup {...userFormAgeGroupProps} />
        <Form.DateSelector {...userFormDateSelectorProps} />
        <Form.Checkbox {...userFormTnCProps} />
        <Form.Textarea {...userFormCommentProps} />
        <Form.Toggle {...userFormToggleProps} />
      </div>
      <Button primary type="submit">
        Submit
      </Button>
    </Form>
  ));
