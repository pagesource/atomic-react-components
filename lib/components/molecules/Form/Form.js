// @flow
import React from 'react';
import { Formik, Form as FormWrapper, Field } from 'formik';
import type { Node } from 'react';
import { buildSchema } from '../../../core/utils/buildValidationSchema';
import { noop } from '../../../core/utils/noop';
import Debug from './Debug';
import FieldInput from './FieldInput';
import FieldSelect from './FieldSelect';
import FieldTextarea from './FieldTextarea';
import FieldRadioGroup from './FieldRadioGroup';
import FieldCheckbox from './FieldCheckbox';
import FieldToggle from './FieldToggle';
import FieldFileInput from './FieldFileInput';
import DDFileInput from './DragDropInput';

type Props = {
  className?: string,
  action: string,
  initialValues?: any,
  children: Node,
  handleSubmit: Function,
  debug?: boolean,
  validationSchema?: Object | null,
  validate?: Function | null,
  validateOnChange?: boolean,
  validateOnBlur?: boolean,
  formWrapperProps?: Object,
  yupValidationSchema?: Function | null,
};

class Form extends React.PureComponent<Props> {
  // eslint-disable-next-line
  static defaultProps = {
    debug: false,
    className: '',
    initialValues: null,
    validationSchema: null,
    validateOnChange: true,
    validateOnBlur: true,
    formWrapperProps: {},
    yupValidationSchema: null,
    validate: noop,
  };

  // eslint-disable-next-line react/sort-comp
  getSchema = (
    yupValidationSchema: Function | null = null,
    validationSchema: Object | null = null
  ) => {
    if (yupValidationSchema) return yupValidationSchema;
    return validationSchema ? buildSchema(validationSchema) : null;
  };

  static Checkbox = FieldCheckbox;

  static Field = Field;

  static Input = FieldInput;

  static RadioGroup = FieldRadioGroup;

  static Select = FieldSelect;

  static Textarea = FieldTextarea;

  static Toggle = FieldToggle;

  static FieldFile = FieldFileInput;

  static DDFile = DDFileInput;

  render() {
    const {
      action,
      className,
      handleSubmit,
      initialValues,
      children,
      debug,
      validationSchema,
      validate,
      validateOnChange,
      validateOnBlur,
      formWrapperProps,
      yupValidationSchema,
      ...others
    } = this.props;
    return (
      <Formik
        initialValues={initialValues}
        validationSchema={this.getSchema(yupValidationSchema, validationSchema)}
        validate={validate}
        onSubmit={handleSubmit}
        validateOnChange={validateOnChange}
        validateOnBlur={validateOnBlur}
        {...formWrapperProps}
      >
        {() => (
          <FormWrapper {...others} noValidate>
            {children}
            {debug && <Debug />}
          </FormWrapper>
        )}
      </Formik>
    );
  }
}

export default Form;
