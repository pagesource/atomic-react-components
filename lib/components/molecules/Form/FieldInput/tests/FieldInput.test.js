import React from 'react';
import Form from '../../Form';
import { shallowWithTheme, mountWithTheme } from '../../../../../../__mock__/themeMock';
import { FieldInputVanilla } from '../FieldInput';

describe('FieldInput Component', () => {
  let FieldInput = '';

  test('should render correctly', () => {
    FieldInput = shallowWithTheme(<FieldInputVanilla />);
    expect(FieldInput).toMatchSnapshot();
  });

  test('should render label correctly', () => {
    const props = {
      label: 'Enter username',
      name: 'username',
      type: 'text',
      validationInstructions: 'Enter valid username',
    };
    FieldInput = mountWithTheme(
      <Form>
        <FieldInputVanilla {...props} />
      </Form>
    );
    console.log('TCL: FieldInput', FieldInput);

    expect(FieldInput.find('label').text()).toEqual('Enter username');
  });

  test('should render Input component with proper type', () => {
    const props = {
      name: 'email',
      type: 'email',
      id: 'email',
    };
    FieldInput = mountWithTheme(
      <Form>
        <FieldInputVanilla {...props} />
      </Form>
    );
    expect(FieldInput.find('input')).toHaveLength(1);
  });
});
