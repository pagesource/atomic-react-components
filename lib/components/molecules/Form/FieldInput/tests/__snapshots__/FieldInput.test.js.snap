// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldInput Component should render correctly 1`] = `
<div
  className=""
>
  <Label
    className=""
    elementLocator=""
  />
  <Field
    type="text"
  >
    <Component />
  </Field>
  <FormikConnect(ErrorMessageImpl)
    component={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "baseHash": 524329165,
          "componentId": "FieldError__StyledFieldError-sc-1b3h259-0",
          "isStatic": false,
          "rules": Array [
            [Function],
            ";",
            [Function],
            ";text-align:left;margin:0;",
            [Function],
            ";",
            ";",
          ],
          "staticRulesId": "",
        },
        "displayName": "FieldError__StyledFieldError",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "FieldError__StyledFieldError-sc-1b3h259-0",
        "target": [Function],
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
  />
</div>
`;
